pkg/tracing/tracing.go

```diff
@@ -0,0 +1,128 @@
+
+		if caCert != "" {
+			data, err := os.ReadFile(caCert)
+			if err != nil {
+				klog.ErrorS(err, "failed to extract ca certificate required to generate the transport credentials")
+				return nil, err
+			}
+			pool := x509.NewCertPool()
+			if !pool.AppendCertsFromPEM(data) {
+				klog.Error("failed to create cert pool using the ca certificate provided for generating the transport credentials")
+				return nil, err
+			}
+			klog.Info("Enabling trace GRPC client in secure TLS mode")
+			opts = append(opts, otlptracegrpc.WithTLSCredentials(credentials.NewClientTLSFromCert(pool, "")))
+		} else {
+			klog.Info("Enabling trace GRPC client in insecure mode")
+			opts = append(opts, otlptracegrpc.WithInsecure())
+		}
+
+		client := otlptracegrpc.NewClient(opts...)
+
+		exporter, err := otlptrace.New(ctx, client)
+		if err != nil {
+			klog.ErrorS(err, "failed to create an instance of the trace exporter")
+			return nil, err
+		}
+		if name == "" {
+			klog.V(5).InfoS("no name provided, using default service name for tracing", "name", DefaultServiceName)
+			name = DefaultServiceName
+		}
+		resourceOpts := []sdkresource.Option{sdkresource.WithAttributes(semconv.ServiceNameKey.String(name)), sdkresource.WithSchemaURL(semconv.SchemaURL)}
+		if namespace != "" {
+			klog.V(5).InfoS("no namespace provided, using default namespace")
+			resourceOpts = append(resourceOpts, sdkresource.WithAttributes(semconv.ServiceNamespaceKey.String(namespace)))
+		}
+		resource, err := sdkresource.New(ctx, resourceOpts...)
+		if err != nil {
+			klog.ErrorS(err, "failed to create traceable resource")
+			return nil, err
+		}
+
+		spanProcessor := sdktrace.NewBatchSpanProcessor(exporter)
+		provider = sdktrace.NewTracerProvider(
+			sdktrace.WithSampler(sdktrace.AlwaysSample()),
+			sdktrace.WithSpanProcessor(spanProcessor),
+			sdktrace.WithResource(resource),
+		)
+	} else {
+		klog.Info("Did not find a trace collector endpoint defined. Switching to NoopTraceProvider")
+		provider = trace.NewNoopTracerProvider()
+	}
+
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetTracerProvider(provider)
+	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
+		klog.ErrorS(err, "got error from opentelemetry")
+	}))
+	klog.Info("Successfully setup trace provider")
+	return
+}
+
+// StartSpan starts a new span with the given operation name and attributes.
+func StartSpan(ctx context.Context, tracerName, operationName string) (context.Context, trace.Span, func()) {
+	return StartSpanWithOptions(ctx, tracerName, []trace.TracerOption{}, operationName, []attribute.KeyValue{})
+}
+
+func StartSpanWithOptions(ctx context.Context, tracerName string, traceOpts []trace.TracerOption, operationName string, attributes []attribute.KeyValue) (context.Context, trace.Span, func()) {
+	spanCtx, span := otel.Tracer(tracerName, traceOpts...).Start(ctx, operationName, trace.WithAttributes(attributes...))
+	return spanCtx, span, func() {
+		span.End()
+	}
+}
```

Comment:
are these functions necessary? They seem to just be wrapping the otel functions

Reply:
Actually, it's not. But pretty lifesaver since these funcs prevents code-duplication during the implementation. And provides more readable structure since we abstract some of the unnecessary parameters and function calls (i.e., Start(), WithAttributes(), etc.)

Reply:
This would make more sense if there was something we intentionally wanted to inject on every span, but besides that it's not really recommended to add your own wrapper package for otel

Reply:
I am fine with moving this out and using the native calls directly. Sure it won't be that much of an extra duplication.

Reply:
cc @harshanarayana